Exception filters

Nest에는 application에서 처리되지 않은 모든 예외를 처리하는 예외 built-in exceptions layer이 있음
application 코드에서 예외가 처리되지 않으면 이 계층이 예외를 포착하고, 자동으로 사용자에게 적합한 응답을 보냄

기본적으로 이 작업은 HttpException 유형의 예외(및 해당 하위 클래스)를 처리하는 built-in global exception filter에 의해 수행됨 
예외가 인식되지 않는 경우(HttpException 또는 HttpException에서 상속되는 클래스가 아님) 기본 예외 필터는 다음과 같은 기본 JSON 응답을 생성함

{
  "statusCode": 500,
  "message": "Internal server error"
}

HINT
The global exception filter partially supports the http-errors library. Basically, any thrown exception containing the statusCode and message property will be properly populated and send back as a response (instead of the default InternalServerErrorException for unrecognized exceptions).

Nest는 @nestjs/common 패키지의 내장 HttpException class를 제공함. 
일반적인 HTTP REST/GraphQL API 기반 applications의 경우 특정 오류 조건이 발생할 때 standard HTTP 응답 개체를 보내는 것이 좋음.

@Get()
async findAll() {
  throw new HttpException('Forbidden', HttpStatus.FORBIDDEN);
}

reponse
{
  "statusCode": 403,
  "message": "Forbidden"
}

HINT
HttpStatus는 @nestjs/common package의 helper enum임

HttpException constructor는 응답을 결정하는 두 개의 필수 인수를 사용함
response 인수는 JSON response body을 정의함. 문자열 또는 객체일 수 있음.
status 인수는 HTTP 상태 코드를 정의함.

기본적으로 JSON response body에는 다음 두 가지 속성이 있음.
statusCode: status 인수에 제공된 HTTP 상태 코드가 기본값임.
메시지: 상태를 기준으로 HTTP 오류에 대한 간단한 설명


JSON response body의 메시지 부분만 재정의하려면 response 인수에 문자열을 입력하면 됨. 
전체 JSON response body을 재정의하려면 response body에 개체를 전달. Nest는 객체를 직렬화하고 JSON response body으로 반환함.
두 번째 constructor 인수인 status는 유효한 HTTP 상태 코드여야 함. @nestjs/common에서 가져온 HttpStatus enum를 사용하는 것이 좋음.

다음은 전체 응답 본문을 재정의하는 예.

@Get()
async findAll() {
  throw new HttpException({
    status: HttpStatus.FORBIDDEN,
    error: 'This is a custom message',
  }, HttpStatus.FORBIDDEN);
}

reponse
{
  "statusCode": 403,
  "message": "Forbidden"
}


Custom exceptions
대부분의 경우 custom exceptions를 작성할 필요가 없으며 다음 섹션에서 설명하는 대로 기본 제공 Nest HTTP exception를 사용할 수 있음. 
customized exceptions를 만들어야 하는 경우 customized exceptions가 기본 HttpException 클래스에서 상속되는 자체 예외 계층을 만드는 것이 좋음. 이 접근 방식을 통해 Nest는 사용자의 예외를 인식하고 오류 응답을 자동으로 처리함. 

//forbidden.exception.ts
export class ForbiddenException extends HttpException {
  constructor() {
    super('Forbidden', HttpStatus.FORBIDDEN);
  }
}

ForbiddenException은 기본 HttpException을 확장하므로 built-in exception handler와 원활하게 작동하므로 findAll() 메서드에서 사용할 수 있음

@Get()
async findAll() {
  throw new ForbiddenException();
}


Built-in HTTP exceptions
Nest는 기본 HttpException에서 상속되는 standard exceptions 집합을 제공함. 이것들은 @nestjs/common 패키지에서 노출되며 가장 일반적인 HTTP 예외의 대부분을 나타냄.

- BadRequestException
- UnauthorizedException
- NotFoundException
- ForbiddenException
- NotAcceptableException
- RequestTimeoutException
- ConflictException
- GoneException
- HttpVersionNotSupportedException
- PayloadTooLargeException
- UnsupportedMediaTypeException
- UnprocessableEntityException
- InternalServerErrorException
- NotImplementedException
- ImATeapotException
- MethodNotAllowedException
- BadGatewayException
- ServiceUnavailableException
- GatewayTimeoutException
- PreconditionFailedException


Exception filters

base (built-in) exception filter는 자동으로 많은 경우를 처리할 수 있지만 예외 계층을 완전히 제어해야 할 수도 있음. 
예를 들어 logging을 추가하거나 dynamic factors에 따라 다른 JSON 스키마를 사용할 수 있음. 
예외 필터는 바로 이러한 목적으로 설계되었음. 이를 통해 정확한 제어 흐름과 클라이언트로 반송되는 응답 내용을 제어할 수 있음.
HttpException class의 인스턴스인 예외를 캡처하고 이에 대한 custom response logic을 구현하는 것을 담당하는 예외 필터를 생성해 보겠음.
이를 위해 기본 플랫폼 Request and Response 객체에 액세스해야 함. 원래 URL을 추출하여 logging 정보에 포함할 수 있도록 Request 개체에 액세스함. 
Response 객체를 사용하여 response.json() 메서드를 사용하여 전송되는 응답을 직접 제어함.

http-exception.filter.ts

import { ExceptionFilter, Catch, ArgumentsHost, HttpException } from '@nestjs/common';
import { Request, Response } from 'express';

@Catch(HttpException)
export class HttpExceptionFilter implements ExceptionFilter {
  catch(exception: HttpException, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<Response>();
    const request = ctx.getRequest<Request>();
    const status = exception.getStatus();

    response
      .status(status)
      .json({
        statusCode: status,
        timestamp: new Date().toISOString(),
        path: request.url,
      });
  }
}

HINT
모든 exception filters는 generic ExceptionFilter<T> interface를 구현해야 함. 이렇게 하려면 지정된 signature을 사용하여 catch(exception: T, host: ArgumentsHost) method를 제공해야 함. T는 exception 유형을 나타냄.

@Catch(HttpException) decorator는 필요한 메타데이터를 exception filter에 바인딩하여 이 특정 필터가 HttpException 유형의 예외를 찾고 있지만 다른 것은 없음을 Nest에 알림. 
@Catch() decorator는 single parameter 또는 쉼표로 구분된 list를 사용할 수 있음. 이렇게 하면 여러 가지 예외 유형에 대한 필터를 한 번에 설정할 수 있음


Arguments host

catch() method의 매개 변수
exception parameter는 현재 처리 중인 예외 객체임. host parameter는 ArgumentsHost object임.

* ArgumentsHost는 실행 컨텍스트 장에서 자세히 살펴볼 강력한 유틸리티 객체임. 
이 코드 샘플에서는 original request handler(예외가 발생한 controller에서)로 전달되는 요청 및 응답 객체에 대한 참조를 얻기 위해 사용함. 이 코드 샘플에서는 ArgumentsHost에서 원하는 요청 및 응답 개체를 가져올 인수에 대한 helper methods을 사용했음.

* 이러한 this level of abstraction의 이유는 ArgumentsHost 함수가 모든 컨텍스트에서 작동하기 때문임(예: 현재 작업 중인 HTTP 서버 context뿐만 아니라 Microservices 및 WebSockets)
실행 컨텍스트 챕터에서는 ArgumentsHost와 이것의 helper functions들이 실행 컨텍스트를 위해 underlying arguments에 어떻게 접근할 수 있는 지에 대해 알아보겠음  
이를 통해 모든 컨텍스트에서 작동하는 일반적인 예외 필터를 작성할 수 있음.

